include/fi_perf.h

```diff
@@ -0,0 +1,386 @@
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _FI_PERF_H_
+#define _FI_PERF_H_
+
+#if HAVE_CONFIG_H
+#  include <config.h>
+#endif /* HAVE_CONFIG_H */
+
+#include <inttypes.h>
+#include <stdio.h>
+
+/* Generic logic */
+
+#define FI_PERF_SLOT_COUNT_MAX       (8)
+#define FI_PERF_SLOT_NAME_LENGTH_MAX (256)
+#define FI_PERF_PRIVATE_AREA_SIZE    (64)
+
+struct fi_perf_slot {
+	uint64_t begin;
+	uint64_t sum;
+	uint64_t count;
+	uint64_t is_active;
+};
+
+struct fi_perf_handle {
+	struct fi_perf_slot slots[FI_PERF_SLOT_COUNT_MAX];
+	char slot_names[FI_PERF_SLOT_COUNT_MAX][FI_PERF_SLOT_NAME_LENGTH_MAX];
+	int count;
+	uint8_t private[FI_PERF_PRIVATE_AREA_SIZE];
+};
+
+#ifdef FI_PERF_RDPMC
+
+/*
+ * The following is a modification/adaptation of a code from Andi Kleen
+ * pmu-tools project:
+ *  https://github.com/andikleen/pmu-tools/blob/master/jevents/rdpmc.c
+*/
+
+/*
+ * Copyright (c) 2012,2013 Intel Corporation
+ * Author: Andi Kleen
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that: (1) source code distributions
+ * retain the above copyright notice and this paragraph in its entirety, (2)
+ * distributions including binary code include the above copyright notice and
+ * this paragraph in its entirety in the documentation or other materials
+ * provided with the distribution
+ *
+ * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED
+ * WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+/* RDPMC specific logic */
+
+#include <unistd.h>
+#include <stdio.h>
+#include <sys/mman.h>
+#include <sys/fcntl.h>
+#include <linux/perf_event.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <asm/unistd.h>
+
+struct fi_perf_rdpmc_context {
+	int fd;
+	struct perf_event_mmap_page *buf;
+	uint32_t type;
+	uint32_t config;
+};
+
+#if defined(__ICC) || defined(__INTEL_COMPILER)
+#include "immintrin.h"
+#endif
+
+#define fi_perf_rdpmc_rmb() asm volatile("" ::: "memory")
+
+/* Default configuration */
+
+#define FI_PERF_RDPMC_DEFAULT_TYPE   (PERF_TYPE_HARDWARE)
+#define FI_PERF_RDPMC_DEFAULT_CONFIG (PERF_COUNT_HW_CPU_CYCLES)
+#define FI_PERF_RDPMC_ALIGNMENT      (4096)
+
+static inline int fi_perf_init(struct fi_perf_handle ** handle_ptr)
+{
+	struct fi_perf_handle * handle = *handle_ptr;
+
+	int result = 0;
+
+	result = posix_memalign((void **)&handle, FI_PERF_RDPMC_ALIGNMENT,
+							sizeof(struct fi_perf_handle));
+	if (result != 0) {
+		/* Error handling */
+		goto exit_err_0;
+	}
+
+	memset(handle, 0, sizeof(struct fi_perf_handle));
+	handle->count = FI_PERF_SLOT_COUNT_MAX;
+
+	struct fi_perf_rdpmc_context * context =
+		(struct fi_perf_rdpmc_context *)handle->private;
+
+	int env_result    = 1;
+	char * env_type   = NULL;
+	char * env_config = NULL;
+
+	env_type = getenv("FI_PERF_RDPMC_TYPE");
```

Comment:
Please use the ofi generic environment variable abstractions, so that these are documented and handled similar to other vars.

Reply:
done